File: sprint2/By lessons/lesson9code1/lesson9code1.js
class Queue {
  constructor(n) {
    this.queue = new Array(n).fill(null);
    this.max_n = n;
    this.head = 0;
    this.tail = 0;
    this.size = 0;
  }

  is_empty() {
    return this.size == 0;
  }

  push(x) {
    if (this.size != this.max_n) {
      this.queue[this.tail] = x;
      this.tail = (this.tail + 1) % this.max_n;
      this.size += 1;
    }
  }

  pop() {
    if (this.is_empty()) {
      return null;
    }
    let x = this.queue[this.head];
    this.queue[this.head] = null;
    this.head = (this.head + 1) % this.max_n;
    this.size -= 1;
    return x;
  }
}

let q = new Queue(8);
q.pop();
console.log(q.queue);  // [null, -1, 0, 11, null, null, null, null]
console.log(q.size);   // 3

q.pop();
q.pop();
q.push(-8);
q.push(7);
q.push(3);
q.push(16);
console.log(q.queue); // [null, null, null, 11, -8, 7, 3, 16]
console.log(q.size); // 5
q.push(12);
console.log(q.queue); // [12, null, null, 11, -8, 7, 3, 16]
console.log(q.size); // 6

File: sprint2/By lessons/lesson9code1/lesson9code1.swift
class Queue<T> {
    private var queue: [T?]
    private var maxN: Int
    private var head: Int
    private var tail: Int
    private var size: Int
    
    init(n: Int) {
        queue = [T?](repeating: nil, count: n)
        maxN = n
        head = 0
        tail = 0
        size = 0
    }
    
    func isEmpty() -> Bool {
        return size == 0
    }
    
    func push(_ x: T) {
        if size != maxN {
            queue[tail] = x
            tail = (tail + 1) % maxN
            size += 1
        }
    }
    
    func pop() -> T? {
        if isEmpty() {
            return nil
        }
        let x = queue[head]
        queue[head] = nil
        head = (head + 1) % maxN
        size -= 1
        return x
    }
    
    func getQueue() -> [T?] {
        return queue
    }
    
    func getSize() -> Int {
        return size
    }
}

var q = Queue<Int>(n: 8)
q.push(42)
q.push(-1)
q.push(0)
q.push(11)
q.pop()

print(q.getQueue()) // [nil, Optional(-1), Optional(0), Optional(11), nil, nil, nil, nil]
print(q.getSize())  // 3

q.pop()
q.pop()
q.push(-8)
q.push(7)
q.push(3)
q.push(16)
print(q.getQueue()) // [nil, nil, nil, Optional(11), Optional(-8), Optional(7), Optional(3), Optional(16)]
print(q.getSize())  // 5
q.push(12)
print(q.getQueue()) // [Optional(12), nil, nil, Optional(11), Optional(-8), Optional(7), Optional(3), Optional(16)]
print(q.getSize())  // 6
